src/collection/backend/in_memory-per_process.cc

```diff
@@ -106,7 +106,7 @@ void InMemoryPerProcess::resolveMultiMatches(const std::string& var,
         }
     } else {
         for (auto &a : *this) {
-            if (a.first.compare(0, var.size(), var) == 0) {
+            if (strncasecmp(a.first.c_str(), var.c_str(), a.first.size()) == 0) {
```

Comment:
Ignoring the change from case-sensitive to case-insensitive (which is the correct fix):

The old code compares the first var.size() characters, the new code compares the first a.first.size() characters. If these two sizes are not the same, this is not equivalent.

Consider var = "tx.paranoia", key = "tx.paranoia_level": the old comparison is true, the new one false.

I don't know why the logic was changed from full comparison to partial comparison in the first place (it looks wrong to me), but with this commit we're changing the logic again.

It would be good to learn about the motivation of the original change (https://github.com/SpiderLabs/ModSecurity/commit/892beb53603a44a3b7cdcbfe83eff9e8c51c9307). One aspect is that it turned a hash lookup into a linear search, so it may well have performance impact.

Reply:
@airween : I agree with you, as I said in my comment. For the purpose of this PR, however, I was just pointing out the change in logic that is not related to case sensitivity, which is the only thing this commit should change.

I've raised the question of comparing only the first N bytes on the original PR.

Reply:
@michaelgranzow-avi : sorry, may be you misunderstood me. I don't understand why is there any str**n**... comparing, instead of "simple" strcmp()/strcasecmp()? Why just need to compare the first N byte, why not the two strings as they are, without any length argument?

Reply:
@airween, I know what you mean and I agree with you. I've opened https://github.com/SpiderLabs/ModSecurity/issues/1818 and added a patch.

Reply:
Proposed fix: https://github.com/SpiderLabs/ModSecurity/pull/1820 - this would make the current PR https://github.com/SpiderLabs/ModSecurity/pull/1810 unnecessary.